fix: tree pagination edge cases (max items 0, descending pattern combos)

traeok opened the pull request:
## Proposed changes

Fixes two edge cases from #3585:

1. Setting `Zowe -> Ds -> Paginate -> Max Items Per Page` to 0 resulted in an error from the paginator. Now, when setting max items to 0, pagination is disabled. This serves as an opt out and is consistent with most APIs for "max items" handling, where 0 is interpreted as "off" / no limit
2. Searching for multiple pattern that happened to be in descending/reverse order, e.g. `trae.*, sys1.*`, resulted in strange behavior of page navigation where items from both patterns were interleaved. Now, when `maxLength` is provided to `listDatasets` (e.g. pagination enabled), we'll sort the patterns alphabetically before passing it to `List.dataSetsMatchingPattern` to ensure proper navigation. This is suitable until server-side sorting is supported, as we sort the resultant list in ZE anyway.

### How to test

- Case 1: Set `Zowe -> Ds -> Paginate -> Max Items Per Page` to 0 and try listing data sets under a profile. An error should no longer occur.
- Case 2: Search for a data set pattern that is in descending order, for example: `B.*, A.*`. Navigating between pages should be smoother compared to what's on `main`.

## Release Notes

Milestone: 3.2.0

Changelog:

No changelog as pagination isn't GA yet

## Types of changes

<!-- What types of changes does your code introduce to Zowe Explorer? Put an `x` in all boxes that apply. -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

<!-- Put an `x` in all boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer. -->

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [x] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

awharn commented:
Please update the description of zowe.ds.paginate to include that a value of 0 disables pagination. Otherwise, this is great!

traeok commented:
> Please update the description of zowe.ds.paginate to include that a value of 0 disables pagination. Otherwise, this is great!

Thanks @awharn, addressed your feedback in 15d2898. I also updated the handling so the session node is refreshed when the page size changes to avoid the edge case you spotted when changing from "1" to "0" as the page size

jace-roell reviewed (dismissed):
LGTM! Thanks @traeok for the quick fixes!

adam-wolfe commented:
Not a significant issue, but If I list data sets with `Datasets per page` = 10, then set it to 0, I see only 10 members until I collapse and re-expand the PDS. It might be better if the tree would refresh (and list all of the members) after changing this setting.

traeok commented:
> Not a significant issue, but If I list data sets with `Datasets per page` = 10, then set it to 0, I see only 10 members until I collapse and re-expand the PDS. It might be better if the tree would refresh (and list all of the members) after changing this setting.

Thanks @adam-wolfe, addressed in 2ac6751

adam-wolfe commented:
> > Not a significant issue, but If I list data sets with `Datasets per page` = 10, then set it to 0, I see only 10 members until I collapse and re-expand the PDS. It might be better if the tree would refresh (and list all of the members) after changing this setting.
> 
> Thanks @adam-wolfe, addressed in [2ac6751](https://github.com/zowe/zowe-explorer-vscode/commit/2ac6751c9c605ed0d446677c8c6fffd825deb8f8)

Fix confirmed, thanks @traeok

zFernand0 reviewed (approved):
LGTM! 😋 
---
Thanks for handling these special cases! 🙏 
And for cleaning up some of the code (empty `else`)😅
